src/vision/status_quo/alan_runs_into_stack_trouble.md

```diff
@@ -0,0 +1,60 @@
+[Barbara]: ../characters/barbara.md
+
+[Alan runs into stack allocation trouble and is able to fix problems]: TODO
+
+### The problem
+
+Alan is working on a networking service written using async Rust and the tokio async runtime. One day, after many changes to his code base, Alan runs his application and hits an error:
+
+```
+$ .\target\debug\application.exe
+thread 'main' has overflowed its stack
+```
+
+Perplexed, Alan sees if anything with his application works by seeing if he can get output when the `--help` flag is passed, but he has no luck:
+
+```
+$ .\target\debug\application.exe --help
+thread 'main' has overflowed its stack
+```
+
+### Searching for the solution
+
+Having really only ever seen stack overflow issues caused by recursive functions, Alan desperately tries to find the source of the bug but searching through the codebase for recursive functions only to find none. Having learned that Rust favors stack allocation over heap allocation (a concept Alan didn't really need to worry about before), he decided to also search for structs that looked large but none appeared in his code.
+
+Confused, Alan reached out to Grace for her advice. She suggested making the stack size larger. Although she wasn't a Windows expert, she remembers hearing that stack sizes on Windows might be smaller than on Linux. After much searching, Alan discovers an option do just that: `RUSTFLAGS = "-C link-args=-Wl,-zstack-size=<size in bytes>"`.
+
+While eventually Alan gets the program to run, the stack size must be set to 4GB before it does! This seems untenable, and Alan goes back to the drawing board.
+
+Alan reaches out to Barbara for her expertise in Rust to see if she has something to suggest. Barbara recommends using `RUSTFLAGS = "-Zprint-type-sizes` to print some type sizes and see if anything jumps out. Barbara noted that if Alan does find a type that stands out, it's usually as easy as putting some boxes in that type to provide some indirection and not have everything be stack allocated. Alan never needs the nightly toolchain, but this option requires it so he installs it using `rustup`. After searching through types, one did stand out as being quite large. Ultimately, this was a red herring, and putting parts of it in `Box`es did not help.
+
+### Finding the solution
+
+After getting no where, Alan went home for the weekend defeated. On Monday, he decided to take another look. One piece of code, stuck out to him: the use of the `select!` macro from the `futures` crate. This macro allowed multiple futures to race against each other, returning the value of the first one to finish. This macro required the futures to be pinned which the docs had shown could be done by using `pin_mut!`. Alan didn't fully grasp what `pin_mut!` was actually doing when he wrote that code. The compiler had complained to him that the futures he was passing to `select!` needed to be pinned, and `pin_mut!` was what he found to make the compiler happy.
```

Comment:
I don't think it belongs in this story, but I'd like to read a story written just about discovering the need to use `select!` and some of the challenges in figuring out what it took to get that code to compile.

Reply:
I can ask the Krustlet folks their experience here, but if I remember correctly, they found `select!` through a google search on how to run two futures concurrently.

Reply:
that would suffice :)

Reply:
I guess the point is there's not a very interesting story there =) makes sense